### PR TITLE
Introductory blog post about q-d library

### PR DESCRIPTION
**Pull request by abailly-iohk**

### CODE REVIEW

File: 2022-09-28-introduce-q-d.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+While the overall state machine appears relatively simple on the surface, the actual protocol is actually quite complex, involving cryptographic signatures and _Plutus Smart Contracts_ to ensure the safety of the protocol. This safety is formally expressed in the paper as _properties_ that are proven against an _Adversary Environment_ whose capabilities are well-defined.
+
+In order to guarantee the implementation provides those safety properties, the Hydra team has developed a diversified palette of testing techniques, including the use of quickcheck-dynamic. While the careful Test-Driven Development approach taken gives reasonable confidence most use cases and issues are covered, hopes are high that such a model is able to explore more corner cases and reveal subtle issues in the protocol implementation.
+
+What was sought after is to be able to define and test Hydra Head security properties against the real implementation. As a first example the team tackled to get a feel of how quickcheck-dynamic could used, here is stated one of the properties from the original paper:
```
Comment (ThomasArts) [inline 2022-09-28-introduce-q-d.md:60]:
```suggestion
What was sought after is to be able to define and test Hydra Head security properties against the real implementation. As a first example the team tackled to get a feel of how quickcheck-dynamic could be used, here one of the properties from the original paper is stated :
```

File: 2022-09-28-introduce-q-d.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      (party, payment) <- forAllQ (nonConflictingTx st)
+      action $ Model.NewTx party payment
+      eventually (ObserveConfirmedTx payment)
+    _ -> pass
+ where
+  nonConflictingTx st = withGenQ (genPayment st) (const [])
+  eventually a = action (Wait 10) >> action a
+```
+
+Note that in order to define this property we have introduced two a "pseudo-actions" in the _Model_, `Wait` and `ObserveConfirmedTx`: Those `Action`s have no effect on the model itself, the former being used to introduce some delay in the context of distributed and asynchronous execution, and the latter serving the purpose of _observing_ the current state of the SUT. An alternative formulation would have been to make `ObserveConfirmedTx` return the set of confirmed transactions and then express the condition as a logic predicate within the `conflictFreeLiveness` property's body.
```
Comment (ThomasArts) [inline 2022-09-28-introduce-q-d.md:189]:
```suggestion
Note that in order to define this property we have introduced two "pseudo-actions" in the _Model_, `Wait` and `ObserveConfirmedTx`: Those `Action`s have no effect on the model itself, the former being used to introduce some delay in the context of distributed and asynchronous execution, and the latter serving the purpose of _observing_ the current state of the SUT. An alternative formulation would have been to make `ObserveConfirmedTx` return the set of confirmed transactions and then express the condition as a logic predicate within the `conflictFreeLiveness` property's body.
```

File: 2022-09-28-introduce-q-d.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
+# Model-Based Testing with QuickCheck
+
+[quickcheck-dynamic](https://github.com/input-output-hk/quickcheck-dynamic) is a library jointly developed by Quviq and IOG, whose purpose is to leverage [QuickCheck](https://hackage.haskell.org/package/QuickCheck) to test stateful programs against a _Model_. In other words, it's a [_Model-Based Testing_](https://en.wikipedia.org/wiki/Model-based_testing) tool. This article wants to be a gentle introduction to the use of quickcheck-dynamic for Model-Based Testing. It describes the overall approach, how the library works, and how it's being applied within IOG to improve the reach of our testing efforts.
+
+## Background
+
+Testing stateful or rather effectful code using QuickCheck is not new: In particular, techniques to test _Monadic_ code with QuickCheck have been introduced in [Claessen & Hughes, 2002](https://dl.acm.org/doi/10.1145/636517.636527). `quickcheck-dynamic` is based on a state-machine approach originally implemented by Quviq in closed-source Erlang version of QuickCheck and put to use to test various systems as explained in John Hughes' [paper](https://publications.lib.chalmers.se/records/fulltext/232550/local_232550.pdf).
```
Comment (ThomasArts) [inline 2022-09-28-introduce-q-d.md:7]:
```suggestion
Testing stateful or rather effectful code using QuickCheck is not new: In particular, techniques to test _Monadic_ code with QuickCheck have been introduced in [Claessen & Hughes, 2002](https://dl.acm.org/doi/10.1145/636517.636527). `quickcheck-dynamic` is based on a state-machine approach originally implemented by Quviq in closed-source [Erlang version of QuickCheck](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.148.6554&rep=rep1&type=pdf) and put to use to test various systems as explained in John Hughes' [paper](https://publications.lib.chalmers.se/records/fulltext/232550/local_232550.pdf).
```

**Review by ch1bo — COMMENTED** [review]
Started review today and have not yet had time to finish fully, but some suggestions already.

File: 2022-09-28-introduce-q-d.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+The Model-Based Testing approach supported by quickcheck-dynamic gives developers the tools to explore the state space in a much more systematic way.
+
+### Hydra
+
+[Hydra](https://hydra.family) is a so-called _Layer 2_ solution for Cardano that aims at increasing the throughput and latency of Cardano transactions by moving most of the traffic out of the main chain (_Layer 1_) and into smaller networks. The _Hydra Head_ protocol is described in great details in a [research paper](https://eprint.iacr.org/2020/299.pdf).
+
+At its core, Hydra is a state machine whose transitions are Layer 1 transactions, as depicted in the following picture:
+
+![Hydra State Machine](https://hydra.family/head-protocol/assets/images/hydra-head-lifecycle-b8449385e9041a214bf8c6e52830de3c.svg)
```
[the remark this answers is not in the document]
Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:52]: I think it's the right level of detail but I agree a proper state-machine could make more sense. There's the picture from the paper but it's ugly and hard to understand, and the one from Miro is probably too detailed and complicated hence why I chose this one

**Comment by abailly-iohk** [comment]
@ch1bo Thank you very much for taking the time to review this!

File: 2022-09-28-introduce-q-d.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+IOG already has had experience with state-machine based testing in the Consensus [Storage layer](https://github.com/input-output-hk/ouroboros-network/blob/nfrisby/old-benchmark-ledger-ops/ouroboros-consensus-test/README.md#L1) using [quickcheck-state-machine](https://hackage.haskell.org/package/quickcheck-state-machine) library, but this was not widespread practice across the various teams.
+
+When IOG started working on the Plutus Smart Contracts language and application framework, Quviq's consultancy was sought after to help test the platform and build tools for future Smart Contract implementors. This effort lead to the development of a custom library for testing smart contracts which in turn leverages quickcheck-dynamic's state-machine model and dynamic logic language.
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:11]: NIT: we built qc-d at the same time as ContractModel.

File: 2022-09-28-introduce-q-d.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+## Use Cases
+
+### Example: Thread Registry
+
+The library comes with a complete example defining a model and reference implementation for a _Thread Registry_. It's inspired by a similar example in Erlang from a couple of papers:
+  * [How well are your requirements tested?](https://publications.lib.chalmers.se/records/fulltext/232552/local_232552.pdf)
+  * and [Understanding Formal Specifications through Good Examples](https://mengwangoxf.github.io/Papers/Erlang18.pdf)
+
+The tests here use IOG's concurrent execution simulator library [io-sim](https://github.com/input-output-hk/io-sim) to speed-up testing.
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:23]: TODO for @MaximilianAlgehed: rebase on the latest version of io-sim to get the merged (?) fix to `threadStatus` in the registry example and make sure the registry example actually passes the tests under io-sim.

File: 2022-09-28-introduce-q-d.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+### Lockstep Testing
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:25]: Does it make sense to talk about these applications after talking about how qc-d works?

File: 2022-09-28-introduce-q-d.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+The quickcheck-lockstep library provides generic implementations for most of the methods of the `StateModel` typeclass and dedicated type-classes to relate the Model and the Implementation.
+
+### Plutus Contracts
+
+Within IOG, the quickcheck-dynamic testing approach was initially applied to provide a testing framework for Smart Contracts developers within the Plutus Application Backend. The Plutus documentation contains a [detailed tutorial](https://plutus-apps.readthedocs.io/en/latest/plutus/tutorials/contract-models.html) on how to model smart contracts and tests them using underlying _Emulator_.
+
+While the _Contract Model_ is a specialised library dedicated to Smart contracts modeling and testing, the underlying principles are similar:
+* Define a `ContractModel` with some state and actions representing the system behaviour,
+* Define a `perform`  function that describes how the model's actions translate to real world calls to a Contract's _endpoints_,
+* then test the contracts implementation using properties written in the _Dynamic Logic_ language provided by the framework.
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:40]: Same comment as above, but this is probably more understandable if it's presented after an example.

File: 2022-09-28-introduce-q-d.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+When IOG started working on the Plutus Smart Contracts language and application framework, Quviq's consultancy was sought after to help test the platform and build tools for future Smart Contract implementors. This effort lead to the development of a custom library for testing smart contracts which in turn leverages quickcheck-dynamic's state-machine model and dynamic logic language.
+
+As quickcheck-dynamic matured and became independent of the Plutus framework, it attracted interest from other teams willing to invest into model-based testing and reuse existing effort. This finally lead to publication of the library as an independent package on [Hackage](https://hackage.haskell.org/package/quickcheck-dynamic) in the hope it will be useful to a wider audience.
+
+## Use Cases
+
+### Example: Thread Registry
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:17]: Maybe we should actually go through this example here and write a full "showcase"? Or is that too much for this blogpost? I worry about a lack of "simple" examples?

File: 2022-09-28-introduce-q-d.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+* A `nextState` (transition) function that evolves the model state according to the `Action`s,
+* Auxiliary function `actionName` to providea  human-readable representation of actions.
+
+The reader is invited to check the [Haddock](https://hackage.haskell.org/package/quickcheck-dynamic-1.1.0/docs/Test-QuickCheck-StateModel.html) documentation of the library for further details.
+
+### Exercising Implementation
+
+A _Model_ alone is somewhat useless if we don't provide a way to relate it to the actual implementation of the system-under-test. quickcheck-dynamic provides the [`RunModel`](https://hackage.haskell.org/package/quickcheck-dynamic-1.1.0/docs/Test-QuickCheck-StateModel.html#t:RunModel) typeclass for this purpose. The most important function to define is `perform` which defines how `StateModel`'s `Action` should be executed against the implementation within some monadic context `m`. Having the actual execution `Monad m` be a parameter of the `RunModel` gives more flexibility to the implementor which is not tied to `IO` for example.
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:130]:
```suggestion
A _Model_ alone is somewhat useless if we don't provide a way to relate it to the actual implementation of the system-under-test. `quickcheck-dynamic` provides the [`RunModel`](https://hackage.haskell.org/package/quickcheck-dynamic-1.1.0/docs/Test-QuickCheck-StateModel.html#t:RunModel) typeclass for this purpose. The most important function to define is `perform` which defines how `StateModel`'s `Action` should be executed against the implementation within some monadic context `m`. Having the actual execution `Monad m` be a parameter of the `RunModel` gives more flexibility to the implementor which is not tied to `IO` for example.
```

File: 2022-09-28-introduce-q-d.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+  { nodes :: Map.Map Party (TestHydraNode Tx m)
+  , logger :: Tracer m (HydraNodeLog Tx)
+  }
+```
+
+The `m` parameter is here kept somewhat unconstrained in order to make it possible to run properties within the `IOSim` monad for faster tests execution. Also note the presence of the `logger` field which is used to capture logging output from all the nodes: Should an error happen or a postcondition fail, we can dump the log of each node which is invaluable to troubleshoot such failures. In general, testing systems in a black-box way emphasises the importance of good logging to provide as much context as possible should issues arise, and using quickcheck-dynamic makes no exception.
+
+### Expressing Properties with Dynamic Logic
+
+Once we have a `StateModel` we can express interesting _properties_ to check against our `RunModel`. [Dynamic Logic](https://github.com/input-output-hk/quickcheck-dynamic/blob/abailly-iohk/link-to-blog-post/quickcheck-dynamic/src/Test/QuickCheck/DynamicLogic.hs) allows one to express properties through monadic expressions relating actions, states and logic predicates.
```
Comment (MaximilianAlgehed) [inline 2022-09-28-introduce-q-d.md:162]: I feel like `RunModel` is a bad name for that class. Bike shedding welcome!

File: 2022-09-28-introduce-q-d.md (hunk 7 above)
Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:25]: I pondered about the ordering of the presentation. It seems to me to make sense to start with concrete use cases, somewhat leaving details out and pointing reader to main resources and detail the library later on

File: 2022-09-28-introduce-q-d.md (hunk 9 above)
Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:17]: I think it would be a separate blog post, perhaps even a paper

Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:17]: I viewed this blog post as mostly introductory but focusing on "real world" use cases, eg. Hydra which is waht I know best :)

File: 2022-09-28-introduce-q-d.md (hunk 11 above)
Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:162]: It kind of makes sense when paired with `StateModel` so 🤷

File: 2022-09-28-introduce-q-d.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+We'll use the latter example to illustrate quickcheck-dynamic's principles and give the reader an intuition on the four steps that need to be defined in order to use it: Defining a test _Model_, stating how the model relates to the _Implementation_, expressing _Properties_ and, last but not least, checking properties.
+
+### Defining a Model
+
+In quickcheck-dynamic, a _Model_ is some type, a representation of the expected state of the system-under-test, for which there exists an instance of the [StateModel class](https://github.com/input-output-hk/quickcheck-dynamic/blob/abailly-iohk/blog-post/quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs#L56) which sets the building blocks needed to generate and validate test sequences.
+
+In the case of Hydra, the Model is a `WorldState` data type that control the Head parties and maintains a `GlobalState` which reflects the expected Head state:
+
+```haskell
+data WorldState = WorldState
```
Comment (ch1bo) [inline 2022-09-28-introduce-q-d.md:84]: Maybe we should call this `IdealWorld` to highlight that it models an ideal world situation? And an alternative (future) model will be `NetworkAdversaryWorld`?

File: 2022-09-28-introduce-q-d.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+The actual monad used is a classical `State` monad whose state maps a `Party` to the corresponding client connection to the Hydra node:
+
+```haskell
+data Nodes m = Nodes
+  { nodes :: Map.Map Party (TestHydraNode Tx m)
+  , logger :: Tracer m (HydraNodeLog Tx)
+  }
+```
+
+The `m` parameter is here kept somewhat unconstrained in order to make it possible to run properties within the `IOSim` monad for faster tests execution. Also note the presence of the `logger` field which is used to capture logging output from all the nodes: Should an error happen or a postcondition fail, we can dump the log of each node which is invaluable to troubleshoot such failures. In general, testing systems in a black-box way emphasises the importance of good logging to provide as much context as possible should issues arise, and using quickcheck-dynamic makes no exception.
```
Comment (ch1bo) [inline 2022-09-28-introduce-q-d.md:158]: I'm not sure about this. The `logger` here is not capturing the logging output. Indeed the `IOSim` monad does it and we can retrieve is using `selectTraceEventsDynamic`. So it actually is NOT required to keep this in the `StateT`, but handing the tracer to `perform`/`runModel` would be sufficient for that.

File: 2022-09-28-introduce-q-d.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+```haskell
+prop_HydraModel :: Actions WorldState -> Property
+prop_HydraModel actions = property $
+  runIOSimProp $ do
+    _ <- runActions runIt actions
+    assert True
+```
+
+In this particular instance from Hydra, we need some additional machinery (the `runIOSimProp` function) to handle the execution of some monadic `PropertyM` into `IOSim` monad, turning it into a `Property`.
```
Comment (ch1bo) [inline 2022-09-28-introduce-q-d.md:207]: And also running the `StateT Nodes` monad transformer. But probably fine to gloss over this.

**Review by ch1bo — APPROVED** [review]
Great article, only some minor comment.

File: 2022-09-28-introduce-q-d.md (hunk 12 above)
Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:84]: That's a good idea.

File: 2022-09-28-introduce-q-d.md (hunk 14 above)
Comment (abailly-iohk) [reply 2022-09-28-introduce-q-d.md:207]: Yeah, it's already quite complicated I guess :)

**Comment by abailly-iohk** [comment]
@MaximilianAlgehed @ch1bo I think I have addressed your comments, perhaps you want to have a quick second look?
